WebSocket Encoder has to be explicitly registered for reflection

**Describe the bug**
Using a custom encoder in the native mode fails with a failure on reflective instantiation.
A user needs to use `@RegisterForReflection` to make it work.

**Expected behavior**
I'd expect all encoders and decoders to be registered for reflection automatically

**Actual behavior**
Application fails to instantiate a [custom encoder](https://github.com/michalszynkiewicz/failing-email-service/blob/master/src/main/java/com/example/JsonEncoder.java#L16
) for a [web socket server](https://github.com/michalszynkiewicz/failing-email-service/blob/master/src/main/java/com/example/Broadcaster.java#L21).

**To Reproduce**
Steps to reproduce the behavior:
1. `git clone https://github.com/michalszynkiewicz/failing-email-service`
2. `mvn clean package -Pnative`
3. go with your browser to `http://localhost:8080`

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
```
Linux t580 5.0.17-200.fc29.x86_64 #1 SMP Mon May 20 15:39:10 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
 - Output of `java -version`: 
```
java version "1.8.0_201"
Java(TM) SE Runtime Environment (build 1.8.0_201-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)
```
 - GraalVM version (if different from Java): 
RC15
 - Quarkus version or git rev: 0.16.0

**Additional context**
The stack trace: 
```
UT026001: Unable to instantiate endpoint: java.lang.RuntimeException: java.lang.InstantiationException: Type `com.example.JsonEncoder` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.
        at io.quarkus.arc.runtime.ArcDeploymentTemplate$DefaultInstanceFactory.create(ArcDeploymentTemplate.java:152)
        at io.quarkus.undertow.runtime.UndertowDeploymentTemplate$6$1.createInstance(UndertowDeploymentTemplate.java:414)
        at io.undertow.websockets.jsr.EncodingFactory.createEncoding(EncodingFactory.java:126)
        at io.undertow.websockets.jsr.EndpointSessionHandler.onConnect(EndpointSessionHandler.java:121)
        at io.undertow.websockets.jsr.JsrWebSocketFilter$1.handleUpgrade(JsrWebSocketFilter.java:165)
        at io.undertow.server.protocol.http.HttpReadListener.exchangeComplete(HttpReadListener.java:385)
        at io.undertow.server.protocol.http.HttpServerConnection.exchangeComplete(HttpServerConnection.java:232)
        at io.undertow.server.HttpServerExchange.invokeExchangeCompleteListeners(HttpServerExchange.java:1279)
        at io.undertow.server.HttpServerExchange.terminateResponse(HttpServerExchange.java:1563)
        at io.undertow.server.Connectors.terminateResponse(Connectors.java:147)
        at io.undertow.server.protocol.http.HttpTransferEncoding$3.handleEvent(HttpTransferEncoding.java:197)
        at io.undertow.server.protocol.http.HttpTransferEncoding$3.handleEvent(HttpTransferEncoding.java:195)
        at io.undertow.conduits.HeadStreamSinkConduit.exitFlush(HeadStreamSinkConduit.java:192)
        at io.undertow.conduits.HeadStreamSinkConduit.flush(HeadStreamSinkConduit.java:133)
        at org.xnio.conduits.ConduitStreamSinkChannel.flush(ConduitStreamSinkChannel.java:162)
        at io.undertow.channels.DetachableStreamSinkChannel.flush(DetachableStreamSinkChannel.java:119)
        at org.xnio.channels.Channels.flushBlocking(Channels.java:63)
        at io.undertow.servlet.spec.ServletOutputStreamImpl.close(ServletOutputStreamImpl.java:618)
        at io.undertow.servlet.spec.HttpServletResponseImpl.closeStreamAndWriter(HttpServletResponseImpl.java:486)
        at io.undertow.servlet.spec.HttpServletResponseImpl.responseDone(HttpServletResponseImpl.java:575)
        at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:351)
        at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
        at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
        at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
        at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
        at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
        at io.quarkus.undertow.runtime.UndertowDeploymentTemplate$7$1$1.call(UndertowDeploymentTemplate.java:469)
        at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
        at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
        at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
        at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
        at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1395)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:32)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:473)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: java.lang.InstantiationException: Type `com.example.JsonEncoder` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.
        at java.lang.Class.newInstance(DynamicHub.java:740)
        at io.quarkus.arc.runtime.ArcDeploymentTemplate$DefaultInstanceFactory.create(ArcDeploymentTemplate.java:144)
        ... 41 more

```

--- a/extensions/camel/camel-servlet/runtime/src/main/java/io/quarkus/camel/servlet/runtime/CamelServletConfig.java
+++ b/extensions/camel/camel-servlet/runtime/src/main/java/io/quarkus/camel/servlet/runtime/CamelServletConfig.java
@@ -3,8 +3,6 @@ package io.quarkus.camel.servlet.runtime;
 import java.util.List;
 import java.util.Map;
 
-import javax.servlet.Servlet;
-
 import io.quarkus.runtime.annotations.ConfigGroup;
 import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigPhase;

--- a/extensions/undertow-websockets/deployment/src/main/java/io/quarkus/undertow/websockets/deployment/UndertowWebsocketProcessor.java
+++ b/extensions/undertow-websockets/deployment/src/main/java/io/quarkus/undertow/websockets/deployment/UndertowWebsocketProcessor.java
@@ -28,9 +28,11 @@ import javax.websocket.server.ServerApplicationConfig;
 import javax.websocket.server.ServerEndpoint;
 
 import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Type;
 
 import io.quarkus.arc.deployment.BeanDefiningAnnotationBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
@@ -124,8 +126,27 @@ public class UndertowWebsocketProcessor {
                 new ReflectiveClassBuildItem(true, false, annotated.toArray(new String[annotated.size()])));
         reflection.produce(new ReflectiveClassBuildItem(false, false, JsrWebSocketFilter.class.getName()));
 
+        final Collection<AnnotationInstance> serverEndpoints = index.getAnnotations(SERVER_ENDPOINT);
+        for (AnnotationInstance endpoint : serverEndpoints) {
+            if (endpoint.target() instanceof ClassInfo) {
+                ClassInfo clazz = (ClassInfo) endpoint.target();
+                if (!Modifier.isAbstract(clazz.flags())) {
+                    registerForReflection(reflection, endpoint.value("encoders"));
+                    registerForReflection(reflection, endpoint.value("decoders"));
+                }
+            }
+        }
+
         return new ServletContextAttributeBuildItem(WebSocketDeploymentInfo.ATTRIBUTE_NAME,
                 template.createDeploymentInfo(annotated, endpoints, config));
+    }
+
+    private void registerForReflection(BuildProducer<ReflectiveClassBuildItem> reflection, AnnotationValue types) {
+        if (types != null && types.asClassArray() != null) {
+            for (Type type : types.asClassArray()) {
+                reflection.produce(new ReflectiveClassBuildItem(false, false, type.name().toString()));
+            }
+        }
     }
 
     @BuildStep

--- a/extensions/undertow-websockets/deployment/src/main/java/io/quarkus/undertow/websockets/deployment/UndertowWebsocketProcessor.java
+++ b/extensions/undertow-websockets/deployment/src/main/java/io/quarkus/undertow/websockets/deployment/UndertowWebsocketProcessor.java
@@ -23,6 +23,8 @@ import java.util.List;
 import java.util.Set;
 
 import javax.websocket.ClientEndpoint;
+import javax.websocket.ClientEndpointConfig;
+import javax.websocket.ContainerProvider;
 import javax.websocket.Endpoint;
 import javax.websocket.server.ServerApplicationConfig;
 import javax.websocket.server.ServerEndpoint;
@@ -43,11 +45,13 @@ import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.substrate.ServiceProviderBuildItem;
 import io.quarkus.runtime.annotations.ConfigRoot;
 import io.quarkus.undertow.deployment.ServletContextAttributeBuildItem;
 import io.quarkus.undertow.deployment.UndertowBuildItem;
 import io.quarkus.undertow.websockets.runtime.UndertowWebsocketTemplate;
 import io.undertow.websockets.jsr.JsrWebSocketFilter;
+import io.undertow.websockets.jsr.UndertowContainerProvider;
 import io.undertow.websockets.jsr.WebSocketDeploymentInfo;
 
 public class UndertowWebsocketProcessor {
@@ -126,8 +130,19 @@ public class UndertowWebsocketProcessor {
                 new ReflectiveClassBuildItem(true, false, annotated.toArray(new String[annotated.size()])));
         reflection.produce(new ReflectiveClassBuildItem(false, false, JsrWebSocketFilter.class.getName()));
 
-        final Collection<AnnotationInstance> serverEndpoints = index.getAnnotations(SERVER_ENDPOINT);
-        for (AnnotationInstance endpoint : serverEndpoints) {
+        registerCodersForReflection(reflection, index.getAnnotations(SERVER_ENDPOINT));
+        registerCodersForReflection(reflection, index.getAnnotations(CLIENT_ENDPOINT));
+
+        reflection.produce(
+                new ReflectiveClassBuildItem(true, true, ClientEndpointConfig.Configurator.class.getName()));
+
+        return new ServletContextAttributeBuildItem(WebSocketDeploymentInfo.ATTRIBUTE_NAME,
+                template.createDeploymentInfo(annotated, endpoints, config));
+    }
+
+    private void registerCodersForReflection(BuildProducer<ReflectiveClassBuildItem> reflection,
+            Collection<AnnotationInstance> endpoints) {
+        for (AnnotationInstance endpoint : endpoints) {
             if (endpoint.target() instanceof ClassInfo) {
                 ClassInfo clazz = (ClassInfo) endpoint.target();
                 if (!Modifier.isAbstract(clazz.flags())) {
@@ -136,15 +151,12 @@ public class UndertowWebsocketProcessor {
                 }
             }
         }
-
-        return new ServletContextAttributeBuildItem(WebSocketDeploymentInfo.ATTRIBUTE_NAME,
-                template.createDeploymentInfo(annotated, endpoints, config));
     }
 
     private void registerForReflection(BuildProducer<ReflectiveClassBuildItem> reflection, AnnotationValue types) {
         if (types != null && types.asClassArray() != null) {
             for (Type type : types.asClassArray()) {
-                reflection.produce(new ReflectiveClassBuildItem(false, false, type.name().toString()));
+                reflection.produce(new ReflectiveClassBuildItem(true, false, type.name().toString()));
             }
         }
     }
@@ -154,6 +166,12 @@ public class UndertowWebsocketProcessor {
     ServiceStartBuildItem setupWorker(UndertowWebsocketTemplate template, UndertowBuildItem undertow) {
         template.setupWorker(undertow.getUndertow());
         return new ServiceStartBuildItem("Websockets");
+    }
+
+    @BuildStep
+    ServiceProviderBuildItem registerContainerProviderService() {
+        return new ServiceProviderBuildItem(ContainerProvider.class.getName(),
+                UndertowContainerProvider.class.getName());
     }
 
     @BuildStep

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/ClientCodingResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/ClientCodingResource.java
@@ -1,0 +1,44 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+
+import javax.websocket.ContainerProvider;
+import javax.websocket.DeploymentException;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+
+@Path("/ws-client-coding-test")
+public class ClientCodingResource {
+
+    @GET
+    public String codeBothWays(@QueryParam("echoServerUri") String echoServerUri) throws IOException, DeploymentException {
+        ContainerProvider.getWebSocketContainer()
+                .connectToServer(CodingClient.class, URI.create(echoServerUri));
+
+        try {
+            Dto response = CodingClient.messageQueue.poll(20, TimeUnit.SECONDS);
+            return response.getContent();
+        } catch (InterruptedException e) {
+            CodingClient.close();
+            return "Failed to get response in time";
+        }
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/ClientDtoDecoder.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/ClientDtoDecoder.java
@@ -1,0 +1,43 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.Reader;
+
+import javax.websocket.Decoder;
+import javax.websocket.EndpointConfig;
+
+public class ClientDtoDecoder implements Decoder.TextStream<Dto> {
+    @Override
+    public Dto decode(Reader reader) throws IOException {
+        try (BufferedReader bufferedReader = new BufferedReader(reader)) {
+            String input = bufferedReader.readLine(); // expecting one line input
+            Dto result = new Dto();
+            result.setContent("[decoded]" + input);
+            return result;
+        }
+    }
+
+    @Override
+    public void init(EndpointConfig config) {
+    }
+
+    @Override
+    public void destroy() {
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/ClientDtoEncoder.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/ClientDtoEncoder.java
@@ -1,0 +1,39 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+import java.io.IOException;
+import java.io.Writer;
+
+import javax.websocket.Encoder;
+import javax.websocket.EndpointConfig;
+
+public class ClientDtoEncoder implements Encoder.TextStream<Dto> {
+    @Override
+    public void encode(Dto object, Writer writer) throws IOException {
+        writer.append("[encoded]" + object.getContent());
+    }
+
+    @Override
+    public void init(EndpointConfig config) {
+
+    }
+
+    @Override
+    public void destroy() {
+
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/CodingClient.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/CodingClient.java
@@ -1,0 +1,59 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import javax.websocket.ClientEndpoint;
+import javax.websocket.OnMessage;
+import javax.websocket.OnOpen;
+import javax.websocket.Session;
+
+@ClientEndpoint(decoders = ClientDtoDecoder.class, encoders = ClientDtoEncoder.class)
+public class CodingClient {
+    private static List<Session> sessions = Collections.synchronizedList(new ArrayList<>());
+
+    static LinkedBlockingDeque<Dto> messageQueue = new LinkedBlockingDeque<>();
+
+    @OnOpen
+    public void onOpen(Session session) {
+        sessions.add(session);
+
+        Dto data = new Dto();
+        data.setContent("initial data");
+        session.getAsyncRemote().sendObject(data);
+    }
+
+    @OnMessage
+    public void onMessage(Dto message) {
+        messageQueue.add(message);
+        close();
+    }
+
+    static void close() {
+        for (Session session : sessions) {
+            try {
+                session.close();
+            } catch (IOException ignored) {
+            }
+        }
+
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/Dto.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/Dto.java
@@ -1,0 +1,35 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+public class Dto {
+    private String content;
+
+    public String getContent() {
+        return content;
+    }
+
+    public void setContent(String content) {
+        this.content = content;
+    }
+
+    @Override
+    public String toString() {
+        return "Dto{" +
+                "content='" + content + '\'' +
+                '}';
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/RecodingSocket.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/RecodingSocket.java
@@ -1,0 +1,29 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+import javax.websocket.OnMessage;
+import javax.websocket.server.ServerEndpoint;
+
+@ServerEndpoint(value = "/recoder", encoders = ServerDtoEncoder.class, decoders = ServerDtoDecoder.class)
+public class RecodingSocket {
+    @OnMessage
+    public Dto recode(Dto input) {
+        Dto result = new Dto();
+        result.setContent("[recoded]" + input.getContent());
+        return result;
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/ServerDtoDecoder.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/ServerDtoDecoder.java
@@ -1,0 +1,41 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+import java.io.Reader;
+
+import javax.json.Json;
+import javax.json.JsonObject;
+import javax.websocket.Decoder;
+import javax.websocket.EndpointConfig;
+
+public class ServerDtoDecoder implements Decoder.TextStream<Dto> {
+    @Override
+    public Dto decode(Reader reader) {
+        JsonObject jsonObject = Json.createReader(reader).readObject();
+        Dto result = new Dto();
+        result.setContent(jsonObject.getString("content"));
+        return result;
+    }
+
+    @Override
+    public void init(EndpointConfig config) {
+    }
+
+    @Override
+    public void destroy() {
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/ServerDtoEncoder.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/ServerDtoEncoder.java
@@ -1,0 +1,42 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.it.websocket;
+
+import java.io.Writer;
+
+import javax.json.Json;
+import javax.json.JsonObject;
+import javax.websocket.Encoder;
+import javax.websocket.EndpointConfig;
+
+public class ServerDtoEncoder implements Encoder.TextStream<Dto> {
+    @Override
+    public void encode(Dto object, Writer writer) {
+        JsonObject jsonObject = Json.createObjectBuilder()
+                .add("content", object.getContent())
+                .build();
+        Json.createWriter(writer)
+                .writeObject(jsonObject);
+    }
+
+    @Override
+    public void init(EndpointConfig config) {
+    }
+
+    @Override
+    public void destroy() {
+    }
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/WebsocketTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/WebsocketTestCase.java
@@ -16,6 +16,8 @@
 
 package io.quarkus.it.main;
 
+import static org.hamcrest.Matchers.is;
+
 import java.net.URI;
 import java.util.concurrent.LinkedBlockingDeque;
 import java.util.concurrent.TimeUnit;
@@ -32,12 +34,16 @@ import org.junit.jupiter.api.Test;
 
 import io.quarkus.test.common.http.TestHTTPResource;
 import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
 
 @QuarkusTest
 public class WebsocketTestCase {
 
     @TestHTTPResource("echo")
-    URI uri;
+    URI echoUri;
+
+    @TestHTTPResource("recoder")
+    URI recoderUri;
 
     @Test
     public void websocketTest() throws Exception {
@@ -54,12 +60,59 @@ public class WebsocketTestCase {
                 });
                 session.getAsyncRemote().sendText("hello");
             }
-        }, ClientEndpointConfig.Builder.create().build(), uri);
+        }, ClientEndpointConfig.Builder.create().build(), echoUri);
 
         try {
             Assertions.assertEquals("hello", message.poll(20, TimeUnit.SECONDS));
         } finally {
             session.close();
         }
+    }
+
+    @Test
+    public void websocketServerEncodingAndDecodingTest() throws Exception {
+        LinkedBlockingDeque<String> message = new LinkedBlockingDeque<>();
+        Session session = ContainerProvider.getWebSocketContainer().connectToServer(new Endpoint() {
+            @Override
+            public void onOpen(Session session, EndpointConfig endpointConfig) {
+                session.addMessageHandler(new MessageHandler.Whole<String>() {
+                    @Override
+                    public void onMessage(String s) {
+                        message.add(s);
+                    }
+                });
+                session.getAsyncRemote().sendText("{\"content\":\"message content\"}");
+            }
+        }, ClientEndpointConfig.Builder.create().build(), recoderUri);
+
+        try {
+            Assertions.assertEquals("{\"content\":\"[recoded]message content\"}", message.poll(20, TimeUnit.SECONDS));
+        } finally {
+            session.close();
+        }
+    }
+
+    /**
+     * In this test:
+     * <ul>
+     * <li>the test sends a request to {@link io.quarkus.it.websocket.ClientCodingResource}</li>
+     * <li>the resource constructs a client that uses an encoder and a decoder and talks with EchoSocket</li>
+     * <li>echo service sends its input back to the resource</li>
+     * <li>the resource sends the result to the test</li>
+     * </ul>
+     */
+    @Test
+    public void websocketClientEncodingAndDecodingTest() {
+        String echoUriWsUri = echoUri.toString().replaceFirst("http", "ws");
+        System.out.println("echo uri: " + echoUriWsUri); // mstod remove
+        RestAssured
+                .given()
+                .queryParam("echoServerUri", echoUriWsUri)
+                .when()
+                .get("/ws-client-coding-test")
+                .then()
+                // "initial-data" is sent from the ClientCodingResource to the EchoService
+                // encoder and decoder prepend it with the [encoded] and [decoded] strings.
+                .body(is("[decoded][encoded]initial data"));
     }
 }

